src/utils/async.ts

```diff
@@ -30,6 +30,57 @@ export class ManualPromise {
     }
 }
 
+/**
+ * WaitGroup tracks the progress of multiple asynchronous tasks, allowing a caller to wait for all tasks to complete.
```

Comment:
When should this be used instead of Promise.all() / Promise.allSettled() ? Docstring should probably answer that.

Reply:
Take a look now :)